### PR TITLE
fix: add settlement timeout and emergency refund for stuck directional positions

### PR DESCRIPTION
**Pull request by pplmaverick**
## Problem

`EventBasedPredictionMarket` resolves via UMA's Optimistic Oracle V2, which is a real
improvement over a single-admin-key oracle — anyone can permissionlessly call
`proposePrice()`, and disputes escalate to the DVM. That part works well.

However, there's no fallback if **nobody ever proposes a price** (e.g. a low-interest
market where the `proposerReward` doesn't attract a proposer, or the underlying question
becomes unresolvable). In that case:

- `receivedSettlementPrice` stays `false` forever.
- `settle()` reverts indefinitely (`require(receivedSettlementPrice, ...)`).
- `redeem()` only works for holders of **matched** Long+Short pairs (1:1 burn). Anyone
  holding a one-sided directional position — which is the normal outcome of trading
  through the AMM — has no exit path at all.

I ran into this while building a live weather prediction market on Arc Testnet
(~55 Arc Testnet transactions across create/lock/settle/claim flows) using a simpler
`onlyOracle`-gated design, and traced the same failure mode back to this reference
implementation while evaluating a migration to it.

## Fix

- Adds `settlementDeadline`, set to `block timestamp of initializeMarket() + SETTLEMENT_TIMEOUT`
  (`72 hours`, currently a fixed constant — open to making this configurable per-market).
- Adds `emergencyRefund(longTokensToRedeem, shortTokensToRedeem)`: callable by any token
  holder once `settlementDeadline` has passed with no settlement price received. Pays out
  at a neutral 0.5/0.5 split per token — the same math the contract already uses for the
  OO's "Undetermined" (`5e17`) outcome — since the true result was never established.

## Known limitation / open question for maintainers

`priceDisputed()` re-requests a price with a fresh timestamp but does **not** reset
`settlementDeadline`. If a dispute lands close to the original deadline, `emergencyRefund()`
could theoretically become callable while a legitimate DVM arbitration is still in flight.
I left this as-is and flagged it with a code comment rather than guessing at the right
behavior (extend the deadline on every dispute? cap the number of extensions?) — wanted to
raise it for discussion rather than bake in an assumption.

Also not included yet: unit tests for the new function and the timeout boundary. Happy to
add a test file (mirroring the existing Hardhat/UMA test setup) if the general approach
looks right to maintainers before I invest in test scaffolding.

## Testing

Not yet covered by automated tests — see note above. Manually reasoned through the
solvency invariant: before any settlement, `longToken.totalSupply() == shortToken.totalSupply()
== collateralToken.balanceOf(address(this))` (since `create()`/`redeem()` always move both
sides in lockstep), so the 0.5/0.5 split in `emergencyRefund()` can't be under-collateralized
in the no-dispute case.

### CODE REVIEW

**Comment by pplmaverick** [comment]
Hi @akelani-circle — wanted to follow up on this PR. With Arc mainnet approaching, any directional prediction market deployed using this template will hit a permanent stuck state if no proposer ever calls proposePrice(). Users holding one-sided positions have no exit path.
Happy to add tests or adjust the implementation if that helps move this forward. Would appreciate a quick review before mainnet goes live.

File: contracts/EventBasedPredictionMarket.sol [hunk 1]
```diff
@@ -199,6 +208,9 @@ contract EventBasedPredictionMarket is Testable {
         requestTimestamp = getCurrentTime();
```
Comment (Cassxbt) [inline contracts/EventBasedPredictionMarket.sol:208]: **[P1] Refresh the deadline for a re-requested price**

`priceDisputed()` creates a fresh oracle request but retains the deadline from the original request. A dispute immediately before expiry therefore leaves a live replacement request subject to an already-expired `emergencyRefund()` window.

I reproduced this against `77d9ee6`: after creating 100 pairs and transferring the Short position to a second holder, I triggered `priceDisputed()` at `settlementDeadline - 1`. At `settlementDeadline + 1`, `emergencyRefund(100, 0)` returns 50 collateral to the Long holder while the replacement request is still pending. If that request then resolves `NO`, the Short holder is owed 100 collateral but only 50 remains, so settlement reverts on the collateral transfer.

Renew the deadline when the fresh request is created:

```solidity
requestTimestamp = getCurrentTime();
settlementDeadline = requestTimestamp + SETTLEMENT_TIMEOUT;
_requestOraclePrice();
```

Please add a regression test for the dispute-at-deadline boundary and verify that the later settlement remains fully collateralized.

Comment (pplmaverick) [reply contracts/EventBasedPredictionMarket.sol:208]: Thanks for the PoC — the race condition is real and the fix is correct in isolation.

One thing worth flagging before we land it: unconditionally resetting `settlementDeadline` on every dispute opens a second attack vector. A malicious actor can repeatedly dispute every incoming proposal, pushing the deadline forward indefinitely and permanently blocking `emergencyRefund()`. That trades a one-shot race (current design) for an unbounded griefing path.

Three options as I see it:

**Option A — Reset once, then lock**
Reset `settlementDeadline` on the first dispute only. Subsequent disputes extend the oracle liveness period but leave the deadline fixed.

```solidity
if (!deadlineExtended) {
    settlementDeadline = getCurrentTime() + SETTLEMENT_TIMEOUT;
    deadlineExtended = true;
}
```

**Option B — Cap total extensions**
Allow up to N resets (e.g. `maxDisputeExtensions = 2`), after which the deadline is frozen regardless of further disputes.

**Option C — Keep current design, document the boundary**
Accept the one-shot race as the lesser risk and strengthen the existing code comment so future maintainers understand the explicit trade-off.

I'd lean toward Option A as the simplest fix that closes both vectors. Happy to implement whichever direction the maintainers prefer before adding the regression test.

**Comment by pplmaverick** [comment]
Hi @akelani-circle — separate from the settlement timeout issue in this PR, I've identified a Critical severity vulnerability in the UMA oracle stack deployment: MockOracleAncillary.pushPrice() has no access control, meaning anyone can unilaterally determine the outcome of any disputed market and drain the losing side's collateral. Full attack chain verified against the installed @uma/core source.

Given the September 16 mainnet launch, wanted to flag this urgently through the right channel. Could you help route this to Circle's security team privately?

**Comment by osr21** [comment]
I verified the mechanics you describe against the actual sources, and they check out — but I think the severity framing needs correcting before this goes anywhere as a "Critical vulnerability report," because what you've found is a deployment-configuration hazard in *this repo's scaffolding*, not a vulnerability in the UMA oracle stack.

**What's true (all confirmed):**
- `MockOracleAncillary.pushPrice()` has no access control — anyone can push a price for any pending request (only guard is `require(queryIndex.isValid, "Can't push prices that haven't been requested")`, i.e. the request must exist).
- `scripts/deploy.ts` deploys it and registers it in the Finder as the `"Oracle"` implementation, so any *disputed* OO request escalates to the mock, where an arbitrary party can decide the outcome.

**Why it isn't a vuln in `@uma/core`:** the contract lives at `data-verification-mechanism/test/MockOracleAncillary.sol` — it's UMA's test double for the DVM, permissionless *by design* so test suites and testnet sandboxes can resolve disputes without running a voting round. Reporting it to UMA (or to Circle as an upstream dependency issue) would be reporting a test fixture for behaving as documented.

**Why it also isn't an exploitable production issue today:** `hardhat.config.ts` defines exactly one network, `arcTestnet`, the deploy script mints faucet-funded `TestnetERC20` collateral, and the collateral at risk is valueless test tokens. Nothing in this repo deploys the mock to a mainnet.

**Where the real, legitimate concern lives — and it's worth filing:** this repo is a *reference template*. Someone cloning it for a mainnet launch could carry the Finder wiring over verbatim and ship a market whose dispute-escalation path is an open `pushPrice()`. That failure mode deserves a public issue (no private routing needed — every fact here is visible in the public deploy script and a public UMA test contract; there's no undisclosed detail to protect). Concrete asks for that issue:

1. **A guard in `deploy.ts`**: refuse to deploy `MockOracleAncillary` (or refuse to register it as `"Oracle"`) when `network.chainId` isn't Arc Testnet's, with an explicit error pointing at this exact risk.
2. **A loud README/SECURITY note**: the mock-DVM wiring is testnet-only; a production deployment needs a real dispute-resolution backend — and note the honest complication that UMA's actual DVM (Voting contract) doesn't exist on Arc, so "just wire the real one" isn't currently an option there. A mainnet design would need either UMA's cross-chain OO infrastructure or a different escalation game entirely. That's a genuine open question for the maintainers, and a better use of their attention than a Critical-severity triage.
3. Optionally, a deploy-summary line that prints *which* oracle the Finder resolves for `"Oracle"`, so anyone auditing a deployment sees the mock immediately.

If after all that you still believe there's an angle with real-fund impact I'm not seeing (e.g. a specific deployed mainnet instance using this wiring), then per both this repo's `SECURITY.md` and the routing discussion we just had on arc-node#269: hackerone.com/circle-bbp, regardless of the scope list — not a maintainer ping in a PR thread. But on the facts as verified, a public issue with the two hardening asks above is the right vehicle, and I'd support it.
